lib/cylc/suite_db_mgr.py

```diff
@@ -522,6 +526,31 @@ def restart_upgrade(self):
             pri_dao = self.get_pri_dao()
             pri_dao.upgrade_pickle_to_json()
 
+        try:
+            self.check_forward_compatibility(pri_dao)
+        except (SuiteServiceFileError, SuiteCylcVersionError) as exc:
+            sys.exit("ERROR: {}".format(exc))
+
         # Vacuum the primary/private database file
         pri_dao.vacuum()
         pri_dao.close()
+
+    def check_forward_compatibility(self, pri_dao):
+        """Raises if the existing suite database is incompatible with the
+        current version of Cylc."""
+        last_run_ver = pri_dao.connect().execute(
+            'SELECT value FROM {} WHERE key == "cylc_version"'.format(
+                self.TABLE_SUITE_PARAMS
+            )
+        ).fetchone()
+        if last_run_ver is None:
+            pri_dao.close()
+            raise SuiteServiceFileError(
+                "Suite database is incompatible with Cylc {}, or "
+                "is corrupted".format(CYLC_VERSION)
+            )
+        last_run_ver = last_run_ver[0]
+        major_ver = last_run_ver.split('.')[0]
```

Comment:
This is comparing a string to an integer:

```suggestion
        major_ver = int(last_run_ver.split('.')[0])
```

With a rather amusing side-effect:

```
ERROR: Suite Cylc version 7.8.8-40-gb96c5 is incompatible with Cylc 7
```

Reply:
I miss mypy!

Reply:
mypy wouldn't help here as in Python 2 comparing a string to an integer is type-safe so that line is perfectly valid and executes fine.

Reply:
I mean for Python 3, the old comments-style type hinting looks like it would be a pain